File: client-sdk/csharp/sdk/src/VertexCache/Sdk/VertexCacheSdk.cs
using System;
using System.IO;
using System.Net.Sockets;
using System.Net.Security;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using VertexCache.Sdk.Core;
using VertexCache.Sdk.Transport;
using VertexCache.Sdk.Crypto;
using VertexCache.Sdk.Protocol;
using VertexCache.Sdk.Results;

namespace VertexCache.Sdk
{
    public class VertexCacheSdk
    {
        private readonly VertexCacheSdkOptions _options;
        private readonly ILogger<VertexCacheSdk>? _logger;

        public VertexCacheSdk(VertexCacheSdkOptions options, ILogger<VertexCacheSdk>? logger = null)
        {
            _options = options;
            _logger = logger;
        }

        public async Task<VCacheResult> RunCommandAsync(string command, string[] args)
        {
            _logger?.LogInformation("🏁 RunCommandAsync invoked: '{Command}' with {ArgCount} args", command, args.Length);

            if (string.IsNullOrWhiteSpace(command))
            {
                _logger?.LogWarning("❌ Empty command rejected before socket creation");
                return VCacheResult.Failure(VCacheErrorCode.InvalidCommand, "Command cannot be empty.");
            }

            string rawCommand = CommandFormatter.Format(command, args);

            if (_options.EnableEncryption)
            {
                try
                {
                    string? rawKey = _options.PublicKey;
                    if (string.IsNullOrWhiteSpace(rawKey))
                        return VCacheResult.Failure(VCacheErrorCode.EncryptionError, "Missing public key for encryption.");

                    _logger?.LogInformation("🔐 Encrypting command: {Raw}", rawCommand);
                    int byteLen = Encoding.UTF8.GetByteCount(rawCommand);
                    _logger?.LogInformation("🔢 Byte length before encryption: {Len}", byteLen);

                    if (byteLen > 245)
                        return VCacheResult.Failure(VCacheErrorCode.EncryptionError, $"Message too long for RSA: {byteLen} bytes");

                    string normalizedKey = EncryptionHelper.NormalizePublicKey(rawKey);
                    rawCommand = EncryptionHelper.Encrypt(rawCommand, normalizedKey);
                }
                catch (Exception ex)
                {
                    _logger?.LogError(ex, "❌ Encryption failed");
                    return VCacheResult.Failure(VCacheErrorCode.EncryptionError, $"Failed to encrypt command: {ex.Message}");
                }
            }

            for (int attempt = 0; attempt <= _options.MaxRetries; attempt++)
            {
                try
                {
                    _logger?.LogInformation("🌐 Connecting to {Host}:{Port} (Attempt {Attempt})", _options.ServerHost, _options.ServerPort, attempt + 1);

                    using var client = new TcpClient();
                    var connectTask = client.ConnectAsync(_options.ServerHost, _options.ServerPort);
                    if (!connectTask.Wait(_options.TimeoutMs))
                        throw new TimeoutException("Connection timed out");

                    using var stream = client.GetStream();
                    Stream ioStream = stream;

                    if (_options.EnableEncryptionTransport)
                    {
                        var ssl = new SslStream(stream, false, (sender, certificate, chain, sslPolicyErrors) =>
                        {
                            if (!_options.EnableVerifyCertificate) return true;
                            if (certificate is null || string.IsNullOrWhiteSpace(_options.CertificatePem)) return false;

                            var expected = X509Certificate2.CreateFromPem(_options.CertificatePem);
                            return certificate.GetCertHashString() == expected.GetCertHashString();
                        });

                        await ssl.AuthenticateAsClientAsync(_options.ServerHost);
                        ioStream = ssl;
                    }

                    using var writer = new StreamWriter(ioStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false)) { AutoFlush = true };
                    using var reader = new StreamReader(ioStream, Encoding.UTF8);

                    if (_options.EnableEncryption)
                    {
                        byte[] encryptedBytes = Convert.FromBase64String(rawCommand);
                        _logger?.LogInformation("📤 Sending {Len} encrypted bytes (raw)", encryptedBytes.Length);
                        await ioStream.WriteAsync(encryptedBytes, 0, encryptedBytes.Length);
                        await ioStream.FlushAsync();
                    }
                    else
                    {
                        _logger?.LogInformation("📤 Sending command (plaintext): {Raw}", rawCommand);
                        await writer.WriteLineAsync(rawCommand);
                    }

                    string? response = await reader.ReadLineAsync();
                    return ProtocolParser.Parse(response);
                }
                catch (Exception ex)
                {
                    _logger?.LogWarning(ex, "Attempt {Attempt} failed", attempt + 1);

                    if (attempt == _options.MaxRetries)
                        return VCacheResult.Failure(VCacheErrorCode.NetworkFailure, $"Failed after {_options.MaxRetries + 1} attempts: {ex.Message}");
                }
            }

            return VCacheResult.Failure(VCacheErrorCode.Unknown, "Unexpected error");
        }
    }
}

File: client-sdk/go/sdk/vertex_cache_sdk.go
package sdk

import (
	"context"
	"crypto/tls"
	"encoding/base64"
	"fmt"
	"strings"
	"vertexcache/sdk/core"
	"vertexcache/sdk/crypto"
	"vertexcache/sdk/protocol"
	"vertexcache/sdk/results"
	"vertexcache/sdk/transport"
)

type VertexCacheSdk struct {
	Options *core.VertexCacheSdkOptions
}

func NewVertexCacheSdk(options *core.VertexCacheSdkOptions) *VertexCacheSdk {
	return &VertexCacheSdk{Options: options}
}

//
// ✅ Simple public helper methods
//

func (sdk *VertexCacheSdk) Ping(ctx context.Context) *results.VCacheResult {
	return sdk.runCommand(ctx, "PING", nil)
}

func (sdk *VertexCacheSdk) Set(ctx context.Context, key, value string) *results.VCacheResult {
	return sdk.runCommand(ctx, "SET", []string{key, value})
}

func (sdk *VertexCacheSdk) Get(ctx context.Context, key string) *results.VCacheResult {
	return sdk.runCommand(ctx, "GET", []string{key})
}

func (sdk *VertexCacheSdk) Del(ctx context.Context, key string) *results.VCacheResult {
	return sdk.runCommand(ctx, "DEL", []string{key})
}

//
// 🔒 Internal engine (was RunCommand)
//

func (sdk *VertexCacheSdk) runCommand(ctx context.Context, command string, args []string) *results.VCacheResult {
	if strings.TrimSpace(command) == "" {
		return results.Failure(results.InvalidCommand, "Command cannot be empty.")
	}

	rawCommand := protocol.FormatCommand(command, args)

	if sdk.Options.EnableEncryption {
		if sdk.Options.PublicKey == "" {
			return results.Failure(results.EncryptionError, "Missing public key for encryption.")
		}

		byteLen := len([]byte(rawCommand))
		if byteLen > 245 {
			return results.Failure(results.EncryptionError, fmt.Sprintf("Message too long for RSA: %d bytes", byteLen))
		}

		normalizedKey := crypto.NormalizePublicKey(sdk.Options.PublicKey)
		encrypted, err := crypto.Encrypt(rawCommand, normalizedKey)
		if err != nil {
			return results.Failure(results.EncryptionError, "Failed to encrypt command: "+err.Error())
		}
		rawCommand = encrypted
	}

	var lastErr error

	for attempt := 0; attempt <= sdk.Options.MaxRetries; attempt++ {
		address := fmt.Sprintf("%s:%d", sdk.Options.ServerHost, sdk.Options.ServerPort)
		var conn transport.VCacheConn
		var err error

		if sdk.Options.EnableEncryptionTransport {
			tlsConfig := &tls.Config{
				ServerName:         sdk.Options.ServerHost,
				InsecureSkipVerify: !sdk.Options.EnableVerifyCertificate,
			}
			if sdk.Options.EnableVerifyCertificate && sdk.Options.CertificatePem != "" {
				cert, err := crypto.LoadCertificateFromString(sdk.Options.CertificatePem)
				if err == nil {
					tlsConfig.RootCAs = crypto.CreateCertPool(cert)
				}
			}
			conn, err = transport.NewTLSConnection(address, tlsConfig)
		} else {
			conn, err = transport.NewTCPConnection(address)
		}

		if err != nil {
			lastErr = err
			continue
		}

		defer conn.Dispose()

		if sdk.Options.EnableEncryption {
			decoded, err := base64.StdEncoding.DecodeString(rawCommand)
			if err != nil {
				return results.Failure(results.EncryptionError, "Failed to decode encrypted command: "+err.Error())
			}
			_, err = conn.SendRaw(decoded)
			if err != nil {
				lastErr = err
				continue
			}
		} else {
			_, err = conn.Send(rawCommand)
			if err != nil {
				lastErr = err
				continue
			}
		}

		line, err := conn.ReadLine()
		if err != nil {
			lastErr = err
			continue
		}

		return protocol.Parse(line)
	}

	if lastErr != nil {
		return results.Failure(results.NetworkFailure, lastErr.Error())
	}

	return results.Failure(results.Unknown, "Unexpected error")
}
